### DISCUSSION TITLE
closed pool

### DISCUSSION THREAD

**Discussion by nexovec** [jackc/pgx#2165] [Q&A] [answered]
Sometimes my pgx Pool closes and all the database accesses return `closed pool` error. Is there a way to error handle this or detect this? Couldn't find anything.

**Comment by nexovec** [comment]
I found mentions of how the pool should never be closed unless you call `.Close()`. I'm also surprised that it's not context aware, it might easily just end up blocking for a very long time.

**Comment by jackc** [comment, marked as answer]
> Sometimes my pgx Pool closes and all the database accesses return closed pool error. Is there a way to error handle this or detect this? Couldn't find anything.

> I found mentions of how the pool should never be closed unless you call .Close().

This is correct. The only way a pgxpool is closed is by calling `Close()`. Any errors encountered using the pool will only affect the underlying connections. I strongly suspect your application is inadvertently closing the pool.

> I'm also surprised that it's not context aware, it might easily just end up blocking for a very long time.

That is correct, it does block. But it's not clear what the alternative would be. What is the state of a pool where `Close` failed / was canceled? It's not in a good state, but it's not closed either.

But most of the time, pools are never closed or only closed at program termination. They should live as long as the program execution does. On the rare occasions that a pool needs to be closed a goroutine can be used to close the pool in a non-blocking way.

**Comment by nexovec** [comment]
> That is correct, it does block. But it's not clear what the alternative would be.

Forcibly shutting it down after context cancels.

I only use `Close()` during graceful shutdown as that one last chance to let connections finish work, which is(should be) always context aware for me. But I understand I can just ram it in a goroutine. Or exit directly.

I assumed it was beneficial to call the `Close()` during shutdown. Now that I think about it, I probably don't need to right?

**Comment by nexovec** [comment]
I seem to have tracked down the original issue and you were totally correct that I was caling `.Close()` (it was a Close on server shutdown but the server can restart after failure thing). I think I will just get rid of the `.Close()` calls entirely for now.

It mistakenly led me to believe that calling `Close()` was not in fact the only way of closing a `Pool`. I don't see appeal any more in hooking this behavior or adding a context parameter.
 
Anyways thanks for help with this.
